sycl/include/CL/__spirv/spirv_ops.hpp

```diff
@@ -193,18 +193,23 @@ __spirv_GroupAll(__spv::Scope Execution, bool Predicate) noexcept;
 __SYCL_CONVERGENT__ extern SYCL_EXTERNAL bool
 __spirv_GroupAny(__spv::Scope Execution, bool Predicate) noexcept;
 
-template <typename dataT>
-__SYCL_CONVERGENT__ extern SYCL_EXTERNAL dataT __spirv_GroupBroadcast(
-    __spv::Scope Execution, dataT Value, size_t LocalId) noexcept;
-
-template <typename dataT>
-__SYCL_CONVERGENT__ extern SYCL_EXTERNAL dataT
-__spirv_GroupBroadcast(__spv::Scope Execution, dataT Value,
-                       __ocl_vec_t<size_t, 2> LocalId) noexcept;
-template <typename dataT>
-__SYCL_CONVERGENT__ extern SYCL_EXTERNAL dataT
-__spirv_GroupBroadcast(__spv::Scope Execution, dataT Value,
-                       __ocl_vec_t<size_t, 3> LocalId) noexcept;
+template <typename dataT, typename idT>
+__SYCL_CONVERGENT__ extern SYCL_EXTERNAL
+    typename std::enable_if<std::is_integral<idT>::value, dataT>::type
+    __spirv_GroupBroadcast(__spv::Scope Execution, dataT Value,
+                           idT LocalId) noexcept;
```

Comment:
Either I think, the good side with the tablegen file is that it is quite strict when it comes to constrain things. Writing in the header speeds up prototyping.

There is another part that can now be more easily done if we move everything to the tablegen: handling extension/capability in the frontend. Using the SPIR-V grammar we can bind builtin to capability and provide a comprehensive error message if one attempt to use a builtin outside the allowed set of capabilities. And some (all ?) SPIR-V Intel's extension are registered in the json, so ultimately, they could benefit from that.

Reply:
We don't use json file with SPIR-V grammar, do we?

Reply:
> We don't use json file with SPIR-V grammar, do we?

I'm not sure what you mean here.

The grammar is a json file https://github.com/KhronosGroup/SPIRV-Headers/blob/master/include/spirv/unified1/spirv.core.grammar.json . That's what I was referring to by saying "json". We don't use any of them (core and extended) ATM (in this project or in the SPIRV-LLVM-Translator AFAIK).

I used them to generate the initial template for the table gen file (it follows more or less the order in that file). But because there is no type information, not much can be extracted for our purpose here.

Reply:
I'm not sure I understand your comment about what "can now be more easily done".

> Using the SPIR-V grammar we can bind builtin to capability and provide a comprehensive error message if one attempt to use a builtin outside the allowed set of capabilities. And some (all ?) SPIR-V Intel's extension are registered in the json, so ultimately, they could benefit from that.

Here you mean that we should re-generate tablegen file each time SPIR-V json grammar file is updated. Right?
Could you add a script you used to generate the tablegen template, please?

Reply:
I rebased to grab #1374 + #1384.  The intrinsics I'd updated disappeared, presumably because they are standard SPIR-V and were already required for work-group functionality.

I don't follow the discussion about the tablegen.  If the latest buildbot fails, I'd appreciate some help in identifying what I need to look at.